File: packages/ts/src/load-extension-entries.ts
// Read a TOML config's `[[dirsql.extension]]` array off disk.

import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve as resolvePath } from "node:path";
import { parse as parseToml } from "smol-toml";

// biome-ignore lint/suspicious/noExplicitAny: TOML root has a dynamic shape.
export type Toml = Record<string, any>;

/** Load a config's `[[dirsql.extension]]` entries with its base directory.
 *
 * `null` when the config is missing, unreadable/malformed, or declares no
 * extension array — the caller leaves such configs to the core.
 */
export function loadExtensionEntries(
  configPath: string,
): { entries: Toml[]; base: string } | null {
  if (!existsSync(configPath)) {
    return null;
  }
  let doc: Toml;
  try {
    doc = parseToml(readFileSync(configPath, "utf8")) as Toml;
  } catch {
    // Leave a malformed config for the core to report.
    return null;
  }
  const dirsql = (doc.dirsql ?? {}) as Toml;
  if (!Array.isArray(dirsql.extension)) {
    return null;
  }
  return { entries: dirsql.extension, base: dirname(resolvePath(configPath)) };
}

File: packages/python/dirsql/load_extension_entries.py
"""Reading one TOML config's ``[[dirsql.extension]]`` array off disk."""

from __future__ import annotations

import os

from .toml_module import _toml


def _load_extension_entries(config_path):
    """Return ``(entries, base_dir)`` for a config's ``[[dirsql.extension]]``.

    ``None`` when the config is missing, unreadable/malformed, or declares no
    extension array -- the caller should leave such configs to the core.
    """
    if not os.path.isfile(config_path):
        return None
    try:
        with open(config_path, "rb") as f:
            doc = _toml.load(f)
    except (OSError, _toml.TOMLDecodeError):
        # Leave a malformed / unreadable config for the core to report.
        return None

    entries = (doc.get("dirsql") or {}).get("extension")
    if not isinstance(entries, list):
        return None
    return entries, os.path.dirname(os.path.abspath(config_path))
